File: APP_Binder_CPP_App/test_client.cpp

#define LOG_TAG "TestService"
//#define LOG_NDEBUG 0

#include <fcntl.h>
#include <sys/prctl.h>
#include <sys/wait.h>
#include <binder/IPCThreadState.h>
#include <binder/ProcessState.h>
#include <binder/IServiceManager.h>
#include <cutils/properties.h>
#include <utils/Log.h>
#include <unistd.h>

#include "IHelloService.h"
#include "IGoodbyeService.h"

using namespace android;

int main(int argc, char **argv) {
	int cnt;

	if (argc < 2) {
        ALOGI("Usage:\n");
        ALOGI("%s <readfile>\n", argv[0]);
        ALOGI("%s <hello|goodbye>\n", argv[0]);
        ALOGI("%s <hello|goodbye> <name>\n", argv[0]);
        return -1;
	}

	sp<ProcessState> proc(ProcessState::self());

	sp<IServiceManager> sm = defaultServiceManager();

	if (strcmp(argv[1], "hello") == 0) {

		sp<IBinder> binder = sm->getService(String16("hello"));

		if (binder == 0) {
		    ALOGI("can't get hello service\n");
			return -1;
		}

		sp<IHelloService> service = interface_cast<IHelloService>(binder);

        if (binder == 0) {
            ALOGI("can't get hello service\n");
            return -1;
        }

		if (argc < 3) {
			service->sayhello();
			ALOGI("client call sayhello");
		} else {
			cnt = service->sayhello_to(argv[2]);
			ALOGI("client call sayhello_to, cnt = %d", cnt);
		}
	} else if (strcmp(argv[1], "readfile") == 0) {

		sp<IBinder> binder = sm->getService(String16("hello"));

		if (binder == 0) {
		    ALOGI("can't get hello service\n");
			return -1;
		}

		sp<IHelloService> service = interface_cast<IHelloService>(binder);

		int fd = service->get_fd();

		ALOGI("client call get_fd = %d", fd);

		char buf[500];
		int len;
		int cnt = 0;

		while (1) {
			len = sprintf(buf, "Hello, test_server, cnt = %d", cnt++);
			write(fd, buf, len);

			len = read(fd, buf, 500);
			buf[len] = '\0';
			ALOGI("%s\n", buf);

			sleep(5);
		}
	} else {

		sp<IBinder> binder = sm->getService(String16("goodbye"));

		if (binder == 0) {
		    ALOGI("can't get goodbye service\n");
			return -1;
		}

		sp<IGoodbyeService> service = interface_cast<IGoodbyeService>(binder);

		if (argc < 3) {
			service->saygoodbye();
			ALOGI("client call saygoodbye");
		} else {
			cnt = service->saygoodbye_to(argv[2]);
			ALOGI("client call saygoodbye_to, cnt = %d", cnt);
		}
	}

	return 0;
}

File: APP_Binder_JAVA_App/TestClient.java
import android.provider.Settings;
import android.util.Slog;
import android.os.ServiceManager;
import android.os.IBinder;

public class TestClient {
    private static final String TAG = TestClient.class.getSimpleName();

    public static void main(String args[]) {
        if (args.length == 0) {
            System.out.println("Usage: need parameter: <hello|goobye> [name]");
            return;
        }

        if (args[0].equals("hello")) {
            IBinder binder = ServiceManager.getService("hello");
            if (binder == null) {
                System.out.println("can not get hello service");
                Slog.i(TAG, "can not get hello service");
                return;
            }

            IHelloService svr = IHelloService.Stub.asInterface(binder);

            if (args.length == 1) {
                try {
                    svr.sayhello();
                    System.out.println("call sayhello");
                    Slog.i(TAG, "call sayhello");
                } catch (Exception e) {
                    e.printStackTrace();
                }
            } else {
                try {
                    int cnt = svr.sayhello_to(args[1]);
                    System.out.println("call sayhello_to " + args[1]);
                    Slog.i(TAG, "call sayhello_to " + args[1] + " : cnt = " + cnt);
                } catch (Exception e) {
                    e.printStackTrace();
                }
            }
        } else if (args[0].equals("goodbye")) {
            IBinder binder = ServiceManager.getService("goodbye");
            if (binder == null) {
                System.out.println("can not get goodbye service");
                Slog.i(TAG, "can not get goodbye service");
                return;
            }

            IGoodbyeService svr = IGoodbyeService.Stub.asInterface(binder);

            if (args.length == 1) {
                try {
                    svr.saygoodbye();
                    System.out.println("call goodbye");
                    Slog.i(TAG, "call goodbye");
                } catch (Exception e) {
                    e.printStackTrace();
                }
            } else {
                try {
                    int cnt = svr.saygoodbye_to(args[1]);
                    System.out.println("call goodbye_to " + args[1]);
                    Slog.i(TAG, "call goodbye_to " + args[1] + " : cnt = " + cnt);
                } catch (Exception e) {
                    e.printStackTrace();
                }
            }
        }
    }
}